### PR TITLE
Feature/new ws extension api, ordering of extensions

### PR DESCRIPTION
**Pull request by dpwspoon**
This does not build, its just for a review with cmebarrow for the ordering

### CODE REVIEW

**Comment by dpwspoon** [comment]
Relevant classes WebSocketExtensionFactory.java, WebSocketExtensionFactoryTest.java,

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/AbstractWsBridgeSession.java [hunk 1]
```diff
@@ -69,22 +70,22 @@
 
     protected ScheduledExecutorService scheduler;
     protected DefaultLoginResult loginResult;
-    protected final ActiveExtensions wsExtensions;
+    protected final ActiveWebSocketExtensions wsExtensions;
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/AbstractWsBridgeSession.java]: Can we keep the simpler name (ActiveExtensions)? It's more readable because shorter, and since the package is transport.ws we already know it's WebSocket extensions we're talking about.

Comment (dpwspoon) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/AbstractWsBridgeSession.java]: For the class? Sure, but BTW it was your todo to rename it to ActiveWebSocketExtensions ;-).  For this field it will probably be removed in the near future.

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/ActiveWebSocketExtensions.java [hunk 2]
```diff
@@ -0,0 +1,297 @@
...
+import org.kaazing.gateway.transport.TypedAttributeKey;
+
+/**
+ * {@link ActiveWebSocketExtensions} maintains a list of WebSocketExtensions that have successfully been negotiated. 
+ * The list that it is constructed with has already properly ordered the extensions as required by the 
+ * {@link WebSocketExtensionSpi}
+ */
+public class ActiveWebSocketExtensions{
+    
+    public static final ActiveWebSocketExtensions EMPTY = new ActiveWebSocketExtensions(new ArrayList<WebSocketExtensionSpi>(0));
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/ActiveWebSocketExtensions.java]: Use Collections.emptyList();

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/AbstractWsBridgeSession.java (hunk 1 above)
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/AbstractWsBridgeSession.java]: Oh oops, must've been before I saw that presentation Todd forwarded a few weeks ago. Let's keep the shorter name.

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/ActiveWebSocketExtensions.java [hunk 3]
```diff
@@ -0,0 +1,297 @@
...
+            }
+        }
+        return null;
+    }
+
+    public boolean hasExtension(Class<? extends WebSocketExtensionSpi> extensionClass) {
+        return getExtension(extensionClass) != null;
+    }
+
+//    public static final ActiveWebSocketExtensions EMPTY = new ActiveWebSocketExtensions(new ArrayList<ExtensionHeader>(0), ExtensionHeader.EndpointKind.SERVER);
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/ActiveWebSocketExtensions.java]: Please remove this commented out code.

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java [hunk 4]
```diff
@@ -19,36 +19,63 @@
...
 import org.kaazing.gateway.resource.address.ws.WsResourceAddress;
 import org.kaazing.gateway.transport.http.HttpAcceptSession;
-import org.kaazing.gateway.transport.ws.extension.ExtensionHeader.EndpointKind;
+import org.kaazing.gateway.transport.ws.extension.WebSocketExtensionFactorySpi.ExtensionOrderCategory;
 
 public final class WebSocketExtensionFactory {
 
     private final Map<String, WebSocketExtensionFactorySpi> factoriesRO;
     private final List<ExtensionHeader> supportedExtensionHeaders;
+    // note: extension names are in the order they would like to be on the pipeline in.
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java]: What does "on the pipeline in" mean? For clarity I suggest rewording to something like the following:
extension names are in the order they would like to be on the pipeline when a frame is being written out to the network

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java [hunk 5]
```diff
@@ -19,36 +19,63 @@
...
+    private final List<String> extensionNames;
 
     private WebSocketExtensionFactory(Map<String, WebSocketExtensionFactorySpi> factoriesRO) {
         this.factoriesRO = factoriesRO;
-        this.supportedExtensionHeaders = toWsExtensions(this.getExtensionNames());
+        // properly order the extension names
+        LinkedList<WebSocketExtensionFactorySpi> orderedExtensions = new LinkedList<>();
+        for (WebSocketExtensionFactorySpi factory : factoriesRO.values()) {
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java]: You should be able to use a Comparator to make this code much simpler.

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java [hunk 6]
```diff
@@ -56,4 +56,33 @@
...
+     * An ordered enum that describes the category that an extension requests to be ordered in.  The enum is ordered by closest
+     * to the network
+     */
+    enum ExtensionOrderCategory{
+        /**
+         * Network, meaning something that transforms the network bytes
+         */
+        NETWORK,
+        /**
+         * Emulation, meaning emulating a frame type that should not be exposed to the application
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java]: Add more clarification here: Extensions which transmit websocket binary or text frames which are not application data.

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java [hunk 7]
```diff
@@ -56,4 +56,33 @@
...
+         */
+        EMULATION,
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java]: Consider renaming to "FRAMING" since this category concerns extension which change the websocket frame flow.

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java [hunk 8]
```diff
@@ -56,4 +56,33 @@
     public abstract WebSocketExtensionSpi negotiate(ExtensionHeader requestedExtension, WsResourceAddress address)
             throws ProtocolException;
 
+    /**
+     * This method allows extensions to specify the order that they would like to be placed on the Filter Chain.
+     *
+     * @return A {@link WebSocketExtensionFactorySpi.ExtensionOrderCategory} that specifies the extension type and where it
+     * should be ordered
+     */
+    public ExtensionOrderCategory orderCategory() {
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java]: Please rename this method to getOrderCategory for consistency with getName

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WsExtensionNegotiationResult.java [hunk 9]
```diff
@@ -37,12 +37,12 @@
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WsExtensionNegotiationResult.java]: Please mark this class as deprecated or remove it

File: transport/ws/src/test/java/org/kaazing/gateway/transport/ws/extension/ActiveWebSocketExtensionsTest.java [hunk 10]
```diff
@@ -0,0 +1,74 @@
...
+        });
+        ActiveWebSocketExtensions activeExtensions = new ActiveWebSocketExtensions(negotiatedExtensions);
+        List<WebSocketExtensionSpi> maintainedList = activeExtensions.asList();
+        for(int i = 0; i < negotiatedExtensions.size(); i++){
+            assertSame(negotiatedExtensions.get(i), maintainedList.get(i));
+        }
+        assertEquals(negotiatedExtensions.size(), maintainedList.size());
+    }
+
+    @Test
```
Comment (cmebarrow) [inline transport/ws/src/test/java/org/kaazing/gateway/transport/ws/extension/ActiveWebSocketExtensionsTest.java]: Remove this line and the next

Comment (cmebarrow) [inline transport/ws/src/test/java/org/kaazing/gateway/transport/ws/extension/ActiveWebSocketExtensionsTest.java]: Need to add tests for the other methods

File: transport/ws/src/test/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactoryTest.java [hunk 11]
```diff
@@ -150,51 +156,152 @@ public void testNegotiateWebSocketExtensionsWithParameter() throws ProtocolExcep
...
-        Assert.assertEquals("foo=2", activeExtensions.asList().get(0).getParameters().get(0).toString());
-        Assert.assertEquals(1, activeExtensions.asList().size());
+        assertSame(webSocketExtensionSpi, ActiveWebSocketExtensions.asList().get(0));
+        assertEquals(1, ActiveWebSocketExtensions.asList().size());
     }
 
     @Test
-    @Ignore("TODO when fix ActiveExtensions and add ordering")
-    public void testNegotiateMultipleWebSocketExtensionsWithParameter() throws ProtocolException {
+    public void testOrderingMechanism() {
```
Comment (cmebarrow) [inline transport/ws/src/test/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactoryTest.java]: Maybe just testOrdering?

File: transport/ws/src/test/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactoryTest.java [hunk 12]
```diff
@@ -150,51 +156,152 @@ public void testNegotiateWebSocketExtensionsWithParameter() throws ProtocolExcep
...
+        WebSocketExtensionFactory.addExtensionInCorrectOrder(perMessageDeflateWebSocketExtensionFactory, orderedExtensions);
+        WebSocketExtensionFactory.addExtensionInCorrectOrder(pingPongWebSocketExtensionFactory, orderedExtensions);
+        WebSocketExtensionFactory.addExtensionInCorrectOrder(fooWebSocketExtensionFactory, orderedExtensions);
+        assertSame(fooWebSocketExtensionFactory, orderedExtensions.get(0));
+        assertSame(pingPongWebSocketExtensionFactory, orderedExtensions.get(1));
+        assertSame(perMessageDeflateWebSocketExtensionFactory, orderedExtensions.get(2));
+    }
+
+    @Test
+    public void testNegotiateMultipleWebSocketExtensionsThatMaintainsOrder() throws ProtocolException {
```
Comment (cmebarrow) [inline transport/ws/src/test/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactoryTest.java]: Please rename along the lines of my preceding comment

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java (hunk 4 above)
Comment (dpwspoon) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java]: To simplify things I have removed the constraint on the ordering of this

**Comment by dpwspoon** [comment]
Ping @cmebarrow, K I believe all the feed back has been applied.   I also did a catch up merge to develop, so the easiest way to view the fixes are viewing the last 2 commits.  Also, I did not change ExtensionOrderCategory names to have FRAMING.  I'm still considering what is best for this so I expect to reach a conclusion when we do some of the actual implementation.

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java [hunk 13]
```diff
@@ -19,36 +19,63 @@
...
 import org.kaazing.gateway.resource.address.ws.WsResourceAddress;
 import org.kaazing.gateway.transport.http.HttpAcceptSession;
-import org.kaazing.gateway.transport.ws.extension.ExtensionHeader.EndpointKind;
+import org.kaazing.gateway.transport.ws.extension.WebSocketExtensionFactorySpi.ExtensionOrderCategory;
 
 public final class WebSocketExtensionFactory {
 
     private final Map<String, WebSocketExtensionFactorySpi> factoriesRO;
-    private final List<ExtensionHeader> supportedExtensionHeaders;
+    private final List<String> extensionNames;
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java]: Consider renaming to orderedExtensionNames

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java [hunk 14]
```diff
@@ -62,7 +89,7 @@ private WebSocketExtensionFactory(Map<String, WebSocketExtensionFactorySpi> fact
      * @throws ProtocolException  If the extension cannot be negotiated. Throwing this exception will result
      *                      in failing the WebSocket connection.
      */
-    public WebSocketExtensionSpi negotiate(ExtensionHeader extension, WsResourceAddress address) throws ProtocolException {
+    public WebSocketExtension negotiate(ExtensionHeader extension, WsResourceAddress address) throws ProtocolException {
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java]: This method does not need to be public I think. You can remove the "public" qualifier, leaving it package private so it can be called by the unit tests

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java (hunk 13 above)
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java]: Actually, I think we can just remove this member (see note on getExtensions)

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java [hunk 15]
```diff
@@ -19,36 +19,63 @@
...
 
     /**
-     * Returns the names of all the supported/discovered extensions.
+     * Returns the names of all the supported/discovered extensions
      *
      * @return Collection of extension names
      */
-    public Collection<String> getExtensionNames() {
-        return factoriesRO.keySet();
+    public List<String> getExtensionNames() {
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactory.java]: I think we can remove this method.

File: transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java [hunk 16]
```diff
@@ -53,7 +54,48 @@
...
+         * Network, meaning something that transforms the network bytes
+         */
+        NETWORK,
+    }
+
+    @Override
+    public int compareTo(WebSocketExtensionFactorySpi o) {
+        List<ExtensionOrderCategory> extensionOrderValues =
+                Arrays.asList(WebSocketExtensionFactorySpi.ExtensionOrderCategory.values());
+        return extensionOrderValues.indexOf(getOrderCategory()) - extensionOrderValues.indexOf(o.getOrderCategory());
```
Comment (cmebarrow) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java]: I think you can just use the enum.ordinal method, so replace the entire method body with
`````
return getOrderCategory().ordinal() - o.getOrderCategory().ordinal();
`````

Comment (dpwspoon) [inline transport/ws/src/main/java/org/kaazing/gateway/transport/ws/extension/WebSocketExtensionFactorySpi.java]: Awesome!
